reframe/core/variables.py

```diff
@@ -34,27 +34,47 @@ class TestVar:
     during class instantiation.
 
     To support injecting attributes into the variable, this class implements a
-    separate dict `__attrs__` where these will be stored.
+    separate dict `_attrs` where these will be stored.
 
     :meta private:
     '''
 
     __slots__ = (
-        'field', '_default_value', 'name', '__attrs__'
+        '_attrs', '_conv_fn', '_default_value', 'field', 'name'
     )
 
     def __init__(self, *args, **kwargs):
         field_type = kwargs.pop('field', fields.TypedField)
         self._default_value = kwargs.pop('value', Undefined)
-
         if not issubclass(field_type, fields.Field):
             raise TypeError(
                 f'field {field_type!r} is not derived from '
                 f'{fields.Field.__qualname__}'
             )
 
+        conv_fn = kwargs.pop('conv', None)
         self.field = field_type(*args, **kwargs)
-        self.__attrs__ = dict()
+        self._attrs = dict()
+
+        def _default_conv(s):
+            errmsg = f'cannot convert {s!r} for setting variable {self.name!r}'
+            if not isinstance(self.field, fields.TypedField):
+                raise TypeError(errmsg)
+
+            for typ in self.field.valid_types:
+                if issubclass(typ, type(None)) and s == 'None':
+                    # Treat `None` specially
+                    return None
+
+                try:
+                    return typ(s)
+                except Exception:
+                    continue
+
+            raise TypeError(errmsg)
+
+        # Set the conversion function
+        self._conv_fn = conv_fn or _default_conv
```

Comment:
The variables just store the default value until it gets injected during class instantiation. This is completely transparent of the value's type, and anything type-related is just down to the descriptor. Conceptually, I think this conversion should not be here.